Code warning for summary.c

```
/home/runner/work/NCEPLIBS-w3emc/NCEPLIBS-w3emc/w3emc/src/summary.c:233:7: warning: implicit declaration of function ‘gettimeofday’ [-Wimplicit-function-declaration]
  233 |   if (gettimeofday (&st, NULL) == -1) {
      |       ^~~~~~~~~~~~
```

--- a/.github/workflows/developer.yml
+++ b/.github/workflows/developer.yml
@@ -74,3 +74,8 @@ jobs:
           w3emc/build/*.html
           w3emc/build/*.css
 
+    - uses: actions/upload-artifact@v2
+      with:
+        name: docs
+        path: |
+          bufr/build/docs/html

--- a/.github/workflows/developer.yml
+++ b/.github/workflows/developer.yml
@@ -49,16 +49,13 @@ jobs:
 
     - name: build-w3emc
       run: |
+        set -x
+        doxygen --version
         cd w3emc
         mkdir build 
         cd build
         cmake -DCMAKE_PREFIX_PATH="~/;~/bacio" -DENABLE_DOCS=ON -DCMAKE_Fortran_FLAGS="-g -fprofile-arcs -ftest-coverage -O0 -fsanitize=address -Wall" -DCMAKE_BUILD_TYPE=Debug ..
         make -j2
-
-    - name: build-docs
-      run: |
-        cd w3emc/build
-        make doc
 
     - name: test-w3emc
       run: |

--- a/docs/Doxyfile.in
+++ b/docs/Doxyfile.in
@@ -1,4 +1,4 @@
-# Doxyfile 1.8.17
+# Doxyfile 1.9.1
 
 # This file describes the settings to be used by the documentation system
 # doxygen (www.doxygen.org) for a project.
@@ -217,6 +217,14 @@ QT_AUTOBRIEF           = YES
 
 MULTILINE_CPP_IS_BRIEF = YES
 
+# By default Python docstrings are displayed as preformatted text and doxygen's
+# special commands cannot be used. By setting PYTHON_DOCSTRING to NO the
+# doxygen's special commands can be used and the contents of the docstring
+# documentation blocks is shown as doxygen documentation.
+# The default value is: YES.
+
+PYTHON_DOCSTRING       = YES
+
 # If the INHERIT_DOCS tag is set to YES then an undocumented member inherits the
 # documentation from any documented member that it re-implements.
 # The default value is: YES.
@@ -252,12 +260,6 @@ TAB_SIZE               = 4
 # a double escape (\\{ and \\})
 
 ALIASES                =
-
-# This tag can be used to specify a number of word-keyword mappings (TCL only).
-# A mapping has the form "name=value". For example adding "class=itcl::class"
-# will allow you to use the command class in the itcl::class meaning.
-
-TCL_SUBST              =
 
 # Set the OPTIMIZE_OUTPUT_FOR_C tag to YES if your project consists of C sources
 # only. Doxygen will then generate output that is more tailored for C. For
@@ -300,18 +302,21 @@ OPTIMIZE_OUTPUT_SLICE  = NO
 # extension. Doxygen has a built-in mapping, but you can override or extend it
 # using this tag. The format is ext=language, where ext is a file extension, and
 # language is one of the parsers supported by doxygen: IDL, Java, JavaScript,
-# Csharp (C#), C, C++, D, PHP, md (Markdown), Objective-C, Python, Slice,
+# Csharp (C#), C, C++, D, PHP, md (Markdown), Objective-C, Python, Slice, VHDL,
 # Fortran (fixed format Fortran: FortranFixed, free formatted Fortran:
 # FortranFree, unknown formatted Fortran: Fortran. In the later case the parser
 # tries to guess whether the code is fixed or free formatted code, this is the
-# default for Fortran type files), VHDL, tcl. For instance to make doxygen treat
-# .inc files as Fortran files (default is PHP), and .f files as C (default is
-# Fortran), use: inc=Fortran f=C.
+# default for Fortran type files). For instance to make doxygen treat .inc files
+# as Fortran files (default is PHP), and .f files as C (default is Fortran),
+# use: inc=Fortran f=C.
 #
 # Note: For files without extension you can use no_extension as a placeholder.
 #
 # Note that for custom extensions you also need to set FILE_PATTERNS otherwise
-# the files are not read by doxygen.
+# the files are not read by doxygen. When specifying no_extension you should add
+# * to the FILE_PATTERNS.
+#
+# Note see also the list of default file extension mappings.
 
 EXTENSION_MAPPING      =
 
@@ -445,6 +450,19 @@ TYPEDEF_HIDES_STRUCT   = NO
 
 LOOKUP_CACHE_SIZE      = 0
 
+# The NUM_PROC_THREADS specifies the number threads doxygen is allowed to use
+# during processing. When set to 0 doxygen will based this on the number of
+# cores available in the system. You can set it explicitly to a value larger
+# than 0 to get more control over the balance between CPU load and processing
+# speed. At this moment only the input processing can be done using multiple
+# threads. Since this is still an experimental feature the default is set to 1,
+# which efficively disables parallel processing. Please report any issues you
+# encounter. Generating dot graphs in parallel is controlled by the
+# DOT_NUM_THREADS setting.
+# Minimum value: 0, maximum value: 32, default value: 1.
+
+NUM_PROC_THREADS       = 1
+
 #---------------------------------------------------------------------------
 # Build related configuration options
 #---------------------------------------------------------------------------
@@ -508,6 +526,13 @@ EXTRACT_LOCAL_METHODS  = NO
 
 EXTRACT_ANON_NSPACES   = YES
 
+# If this flag is set to YES, the name of an unnamed parameter in a declaration
+# will be determined by the corresponding definition. By default unnamed
+# parameters remain unnamed in the output.
+# The default value is: YES.
+
+RESOLVE_UNNAMED_PARAMS = YES
+
 # If the HIDE_UNDOC_MEMBERS tag is set to YES, doxygen will hide all
 # undocumented members inside documented classes or files. If set to NO these
 # members will be included in the various overviews, but no documentation
@@ -545,11 +570,18 @@ HIDE_IN_BODY_DOCS      = NO
 
 INTERNAL_DOCS          = NO
 
-# If the CASE_SENSE_NAMES tag is set to NO then doxygen will only generate file
-# names in lower-case letters. If set to YES, upper-case letters are also
-# allowed. This is useful if you have classes or files whose names only differ
-# in case and if your file system supports case sensitive file names. Windows
-# (including Cygwin) ands Mac users are advised to set this option to NO.
+# With the correct setting of option CASE_SENSE_NAMES doxygen will better be
+# able to match the capabilities of the underlying filesystem. In case the
+# filesystem is case sensitive (i.e. it supports files in the same directory
+# whose names only differ in casing), the option must be set to YES to properly
+# deal with such files in case they appear in the input. For filesystems that
+# are not case sensitive the option should be be set to NO to properly deal with
+# output files written for symbols that only differ in casing, such as for two
+# classes, one named CLASS and the other named Class, and to also support
+# references to files without having to specify the exact matching casing. On
+# Windows (including Cygwin) and MacOS, users should typically set this option
+# to NO, whereas on Linux or other Unix flavors it should typically be set to
+# YES.
 # The default value is: system dependent.
 
 CASE_SENSE_NAMES       = YES
@@ -788,7 +820,10 @@ WARN_IF_DOC_ERROR      = YES
 WARN_NO_PARAMDOC       = YES
 
 # If the WARN_AS_ERROR tag is set to YES then doxygen will immediately stop when
-# a warning is encountered.
+# a warning is encountered. If the WARN_AS_ERROR tag is set to FAIL_ON_WARNINGS
+# then doxygen will continue running as if WARN_AS_ERROR tag is set to NO, but
+# at the end of the doxygen process doxygen will return with a non-zero status.
+# Possible values are: NO, YES and FAIL_ON_WARNINGS.
 # The default value is: NO.
 
 WARN_AS_ERROR          = YES
@@ -825,8 +860,8 @@ INPUT                  = @abs_top_srcdir@/docs/user_guide.md \
 # This tag can be used to specify the character encoding of the source files
 # that doxygen parses. Internally doxygen uses the UTF-8 encoding. Doxygen uses
 # libiconv (or the iconv built into libc) for the transcoding. See the libiconv
-# documentation (see: https://www.gnu.org/software/libiconv/) for the list of
-# possible encodings.
+# documentation (see:
+# https://www.gnu.org/software/libiconv/) for the list of possible encodings.
 # The default value is: UTF-8.
 
 INPUT_ENCODING         = UTF-8
@@ -839,15 +874,22 @@ INPUT_ENCODING         = UTF-8
 # need to set EXTENSION_MAPPING for the extension otherwise the files are not
 # read by doxygen.
 #
+# Note the list of default checked file patterns might differ from the list of
+# default file extension mappings.
+#
 # If left blank the following patterns are tested:*.c, *.cc, *.cxx, *.cpp,
 # *.c++, *.java, *.ii, *.ixx, *.ipp, *.i++, *.inl, *.idl, *.ddl, *.odl, *.h,
 # *.hh, *.hxx, *.hpp, *.h++, *.cs, *.d, *.php, *.php4, *.php5, *.phtml, *.inc,
 # *.m, *.markdown, *.md, *.mm, *.dox (to be provided as doxygen C comment),
-# *.doc (to be provided as doxygen C comment), *.txt (to be provided as doxygen
-# C comment), *.py, *.pyw, *.f90, *.f95, *.f03, *.f08, *.f, *.for, *.tcl, *.vhd,
-# *.vhdl, *.ucf, *.qsf and *.ice.
+# *.py, *.pyw, *.f90, *.f95, *.f03, *.f08, *.f18, *.f, *.for, *.vhd, *.vhdl,
+# *.ucf, *.qsf and *.ice.
 
-FILE_PATTERNS          = *.F90 *.f90 *.f *.F *.c *.h
+FILE_PATTERNS          = *.F90 \
+                         *.f90 \
+                         *.f \
+                         *.F \
+                         *.c \
+                         *.h
 
 # The RECURSIVE tag can be used to specify whether or not subdirectories should
 # be searched for input files as well.
@@ -1059,35 +1101,6 @@ USE_HTAGS              = NO
 
 VERBATIM_HEADERS       = YES
 
-# If the CLANG_ASSISTED_PARSING tag is set to YES then doxygen will use the
-# clang parser (see: http://clang.llvm.org/) for more accurate parsing at the
-# cost of reduced performance. This can be particularly helpful with template
-# rich C++ code for which doxygen's built-in parser lacks the necessary type
-# information.
-# Note: The availability of this option depends on whether or not doxygen was
-# generated with the -Duse_libclang=ON option for CMake.
-# The default value is: NO.
-
-CLANG_ASSISTED_PARSING = NO
-
-# If clang assisted parsing is enabled you can provide the compiler with command
-# line options that you would normally use when invoking the compiler. Note that
-# the include paths will already be set by doxygen for the files and directories
-# specified with INPUT and INCLUDE_PATH.
-# This tag requires that the tag CLANG_ASSISTED_PARSING is set to YES.
-
-CLANG_OPTIONS          =
-
-# If clang assisted parsing is enabled you can provide the clang parser with the
-# path to the compilation database (see:
-# http://clang.llvm.org/docs/HowToSetupToolingForLLVM.html) used when the files
-# were built. This is equivalent to specifying the "-p" option to a clang tool,
-# such as clang-check. These options will then be passed to the parser.
-# Note: The availability of this option depends on whether or not doxygen was
-# generated with the -Duse_libclang=ON option for CMake.
-
-CLANG_DATABASE_PATH    =
-
 #---------------------------------------------------------------------------
 # Configuration options related to the alphabetical class index
 #---------------------------------------------------------------------------
@@ -1098,13 +1111,6 @@ CLANG_DATABASE_PATH    =
 # The default value is: YES.
 
 ALPHABETICAL_INDEX     = NO
-
-# The COLS_IN_ALPHA_INDEX tag can be used to specify the number of columns in
-# which the alphabetical index list will be split.
-# Minimum value: 1, maximum value: 20, default value: 5.
-# This tag requires that the tag ALPHABETICAL_INDEX is set to YES.
-
-COLS_IN_ALPHA_INDEX    = 5
 
 # In case all classes in a project start with a common prefix, all classes will
 # be put under the same header in the alphabetical index. The IGNORE_PREFIX tag
@@ -1276,10 +1282,11 @@ HTML_INDEX_NUM_ENTRIES = 100
 
 # If the GENERATE_DOCSET tag is set to YES, additional index files will be
 # generated that can be used as input for Apple's Xcode 3 integrated development
-# environment (see: https://developer.apple.com/xcode/), introduced with OSX
-# 10.5 (Leopard). To create a documentation set, doxygen will generate a
-# Makefile in the HTML output directory. Running make will produce the docset in
-# that directory and running make install will install the docset in
+# environment (see:
+# https://developer.apple.com/xcode/), introduced with OSX 10.5 (Leopard). To
+# create a documentation set, doxygen will generate a Makefile in the HTML
+# output directory. Running make will produce the docset in that directory and
+# running make install will install the docset in
 # ~/Library/Developer/Shared/Documentation/DocSets so that Xcode will find it at
 # startup. See https://developer.apple.com/library/archive/featuredarticles/Doxy
 # genXcode/_index.html for more information.
@@ -1321,8 +1328,8 @@ DOCSET_PUBLISHER_NAME  = Publisher
 # If the GENERATE_HTMLHELP tag is set to YES then doxygen generates three
 # additional HTML index files: index.hhp, index.hhc, and index.hhk. The
 # index.hhp is a project file that can be read by Microsoft's HTML Help Workshop
-# (see: https://www.microsoft.com/en-us/download/details.aspx?id=21138) on
-# Windows.
+# (see:
+# https://www.microsoft.com/en-us/download/details.aspx?id=21138) on Windows.
 #
 # The HTML Help Workshop contains a compiler that can convert all HTML output
 # generated by doxygen into a single compiled HTML file (.chm). Compiled HTML
@@ -1352,7 +1359,7 @@ CHM_FILE               =
 HHC_LOCATION           =
 
 # The GENERATE_CHI flag controls if a separate .chi index file is generated
-# (YES) or that it should be included in the master .chm file (NO).
+# (YES) or that it should be included in the main .chm file (NO).
 # The default value is: NO.
 # This tag requires that the tag GENERATE_HTMLHELP is set to YES.
 
@@ -1397,7 +1404,8 @@ QCH_FILE               =
 
 # The QHP_NAMESPACE tag specifies the namespace to use when generating Qt Help
 # Project output. For more information please see Qt Help Project / Namespace
-# (see: https://doc.qt.io/archives/qt-4.8/qthelpproject.html#namespace).
+# (see:
+# https://doc.qt.io/archives/qt-4.8/qthelpproject.html#namespace).
 # The default value is: org.doxygen.Project.
 # This tag requires that the tag GENERATE_QHP is set to YES.
 
@@ -1405,8 +1413,8 @@ QHP_NAMESPACE          = org.doxygen.Project
 
 # The QHP_VIRTUAL_FOLDER tag specifies the namespace to use when generating Qt
 # Help Project output. For more information please see Qt Help Project / Virtual
-# Folders (see: https://doc.qt.io/archives/qt-4.8/qthelpproject.html#virtual-
-# folders).
+# Folders (see:
+# https://doc.qt.io/archives/qt-4.8/qthelpproject.html#virtual-folders).
 # The default value is: doc.
 # This tag requires that the tag GENERATE_QHP is set to YES.
 
@@ -1414,16 +1422,16 @@ QHP_VIRTUAL_FOLDER     = doc
 
 # If the QHP_CUST_FILTER_NAME tag is set, it specifies the name of a custom
 # filter to add. For more information please see Qt Help Project / Custom
-# Filters (see: https://doc.qt.io/archives/qt-4.8/qthelpproject.html#custom-
-# filters).
+# Filters (see:
+# https://doc.qt.io/archives/qt-4.8/qthelpproject.html#custom-filters).
 # This tag requires that the tag GENERATE_QHP is set to YES.
 
 QHP_CUST_FILTER_NAME   =
 
 # The QHP_CUST_FILTER_ATTRS tag specifies the list of the attributes of the
 # custom filter to add. For more information please see Qt Help Project / Custom
-# Filters (see: https://doc.qt.io/archives/qt-4.8/qthelpproject.html#custom-
-# filters).
+# Filters (see:
+# https://doc.qt.io/archives/qt-4.8/qthelpproject.html#custom-filters).
 # This tag requires that the tag GENERATE_QHP is set to YES.
 
 QHP_CUST_FILTER_ATTRS  =
@@ -1435,9 +1443,9 @@ QHP_CUST_FILTER_ATTRS  =
 
 QHP_SECT_FILTER_ATTRS  =
 
-# The QHG_LOCATION tag can be used to specify the location of Qt's
-# qhelpgenerator. If non-empty doxygen will try to run qhelpgenerator on the
-# generated .qhp file.
+# The QHG_LOCATION tag can be used to specify the location (absolute path
+# including file name) of Qt's qhelpgenerator. If non-empty doxygen will try to
+# run qhelpgenerator on the generated .qhp file.
 # This tag requires that the tag GENERATE_QHP is set to YES.
 
 QHG_LOCATION           =
@@ -1514,6 +1522,17 @@ TREEVIEW_WIDTH         = 250
 
 EXT_LINKS_IN_WINDOW    = NO
 
+# If the HTML_FORMULA_FORMAT option is set to svg, doxygen will use the pdf2svg
+# tool (see https://github.com/dawbarton/pdf2svg) or inkscape (see
+# https://inkscape.org) to generate formulas as SVG images instead of PNGs for
+# the HTML output. These images will generally look nicer at scaled resolutions.
+# Possible values are: png (the default) and svg (looks nicer but requires the
+# pdf2svg or inkscape tool).
+# The default value is: png.
+# This tag requires that the tag GENERATE_HTML is set to YES.
+
+HTML_FORMULA_FORMAT    = png
+
 # Use this tag to change the font size of LaTeX formulas included as images in
 # the HTML documentation. When you change the font size after a successful
 # doxygen run you need to manually remove any form_*.png images from the HTML
@@ -1553,7 +1572,7 @@ USE_MATHJAX            = NO
 
 # When MathJax is enabled you can set the default output format to be used for
 # the MathJax output. See the MathJax site (see:
-# http://docs.mathjax.org/en/latest/output.html) for more details.
+# http://docs.mathjax.org/en/v2.7-latest/output.html) for more details.
 # Possible values are: HTML-CSS (which is slower, but has the best
 # compatibility), NativeMML (i.e. MathML) and SVG.
 # The default value is: HTML-CSS.
@@ -1569,7 +1588,7 @@ MATHJAX_FORMAT         = HTML-CSS
 # Content Delivery Network so you can quickly see the result without installing
 # MathJax. However, it is strongly recommended to install a local copy of
 # MathJax from https://www.mathjax.org before deployment.
-# The default value is: https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/.
+# The default value is: https://cdn.jsdelivr.net/npm/mathjax@2.
 # This tag requires that the tag USE_MATHJAX is set to YES.
 
 MATHJAX_RELPATH        = http://cdn.mathjax.org/mathjax/latest
@@ -1583,7 +1602,8 @@ MATHJAX_EXTENSIONS     =
 
 # The MATHJAX_CODEFILE tag can be used to specify a file with javascript pieces
 # of code that will be used on startup of the MathJax code. See the MathJax site
-# (see: http://docs.mathjax.org/en/latest/output.html) for more details. For an
+# (see:
+# http://docs.mathjax.org/en/v2.7-latest/output.html) for more details. For an
 # example see the documentation.
 # This tag requires that the tag USE_MATHJAX is set to YES.
 
@@ -1630,7 +1650,8 @@ SERVER_BASED_SEARCH    = NO
 #
 # Doxygen ships with an example indexer (doxyindexer) and search engine
 # (doxysearch.cgi) which are based on the open source search engine library
-# Xapian (see: https://xapian.org/).
+# Xapian (see:
+# https://xapian.org/).
 #
 # See the section "External Indexing and Searching" for details.
 # The default value is: NO.
@@ -1643,8 +1664,9 @@ EXTERNAL_SEARCH        = NO
 #
 # Doxygen ships with an example indexer (doxyindexer) and search engine
 # (doxysearch.cgi) which are based on the open source search engine library
-# Xapian (see: https://xapian.org/). See the section "External Indexing and
-# Searching" for details.
+# Xapian (see:
+# https://xapian.org/). See the section "External Indexing and Searching" for
+# details.
 # This tag requires that the tag SEARCHENGINE is set to YES.
 
 SEARCHENGINE_URL       =
@@ -1808,9 +1830,11 @@ LATEX_EXTRA_FILES      =
 
 PDF_HYPERLINKS         = YES
 
-# If the USE_PDFLATEX tag is set to YES, doxygen will use pdflatex to generate
-# the PDF file directly from the LaTeX files. Set this option to YES, to get a
-# higher quality PDF documentation.
+# If the USE_PDFLATEX tag is set to YES, doxygen will use the engine as
+# specified with LATEX_CMD_NAME to generate the PDF file directly from the LaTeX
+# files. Set this option to YES, to get a higher quality PDF documentation.
+#
+# See also section LATEX_CMD_NAME for selecting the engine.
 # The default value is: YES.
 # This tag requires that the tag GENERATE_LATEX is set to YES.
 
@@ -2242,7 +2266,7 @@ HIDE_UNDOC_RELATIONS   = YES
 # http://www.graphviz.org/), a graph visualization toolkit from AT&T and Lucent
 # Bell Labs. The other options in this section have no effect if this option is
 # set to NO
-# The default value is: YES.
+# The default value is: NO.
 
 HAVE_DOT               = NO
 
@@ -2321,9 +2345,31 @@ UML_LOOK               = NO
 # but if the number exceeds 15, the total amount of fields shown is limited to
 # 10.
 # Minimum value: 0, maximum value: 100, default value: 10.
-# This tag requires that the tag HAVE_DOT is set to YES.
+# This tag requires that the tag UML_LOOK is set to YES.
 
 UML_LIMIT_NUM_FIELDS   = 10
+
+# If the DOT_UML_DETAILS tag is set to NO, doxygen will show attributes and
+# methods without types and arguments in the UML graphs. If the DOT_UML_DETAILS
+# tag is set to YES, doxygen will add type and arguments for attributes and
+# methods in the UML graphs. If the DOT_UML_DETAILS tag is set to NONE, doxygen
+# will not generate fields with class member information in the UML graphs. The
+# class diagrams will look similar to the default class diagrams but using UML
+# notation for the relationships.
+# Possible values are: NO, YES and NONE.
+# The default value is: NO.
+# This tag requires that the tag UML_LOOK is set to YES.
+
+DOT_UML_DETAILS        = NO
+
+# The DOT_WRAP_THRESHOLD tag can be used to set the maximum number of characters
+# to display on a single line. If the actual line length exceeds this threshold
+# significantly it will wrapped across multiple lines. Some heuristics are apply
+# to avoid ugly line breaks.
+# Minimum value: 0, maximum value: 1000, default value: 17.
+# This tag requires that the tag HAVE_DOT is set to YES.
+
+DOT_WRAP_THRESHOLD     = 17
 
 # If the TEMPLATE_RELATIONS tag is set to YES then the inheritance and
 # collaboration graphs will show the relations between templates and their
@@ -2398,9 +2444,7 @@ DIRECTORY_GRAPH        = YES
 # Note: If you choose svg you need to set HTML_FILE_EXTENSION to xhtml in order
 # to make the SVG files visible in IE 9+ (other browsers do not have this
 # requirement).
-# Possible values are: png, png:cairo, png:cairo:cairo, png:cairo:gd, png:gd,
-# png:gd:gd, jpg, jpg:cairo, jpg:cairo:gd, jpg:gd, jpg:gd:gd, gif, gif:cairo,
-# gif:cairo:gd, gif:gd, gif:gd:gd, svg, png:gd, png:gd:gd, png:cairo,
+# Possible values are: png, jpg, gif, svg, png:gd, png:gd:gd, png:cairo,
 # png:cairo:gd, png:cairo:cairo, png:cairo:gdiplus, png:gdiplus and
 # png:gdiplus:gdiplus.
 # The default value is: png.
@@ -2516,9 +2560,11 @@ DOT_MULTI_TARGETS      = NO
 
 GENERATE_LEGEND        = YES
 
-# If the DOT_CLEANUP tag is set to YES, doxygen will remove the intermediate dot
+# If the DOT_CLEANUP tag is set to YES, doxygen will remove the intermediate
 # files that are used to generate the various graphs.
+#
+# Note: This setting is not only used for dot files but also for msc and
+# plantuml temporary files.
 # The default value is: YES.
-# This tag requires that the tag HAVE_DOT is set to YES.
 
 DOT_CLEANUP            = YES

--- a/docs/Doxyfile.in
+++ b/docs/Doxyfile.in
@@ -847,10 +847,7 @@ INPUT_ENCODING         = UTF-8
 # C comment), *.py, *.pyw, *.f90, *.f95, *.f03, *.f08, *.f, *.for, *.tcl, *.vhd,
 # *.vhdl, *.ucf, *.qsf and *.ice.
 
-FILE_PATTERNS          = *.F90 \
-                         *.f90 \
-                         *.f \
-                         *.F
+FILE_PATTERNS          = *.F90 *.f90 *.f *.F *.c *.h
 
 # The RECURSIVE tag can be used to specify whether or not subdirectories should
 # be searched for input files as well.

--- a/src/summary.c
+++ b/src/summary.c
@@ -1,38 +1,41 @@
-/***************************************************************
-
-This code will make a system call to return various
-useful parameters.  When subroutine summary is called, a list
-of system resource statistics is printed to stdout.
-
-Users need to place a call to start() at the beginning of the
-section of code to be "measured" and a call to summary() at the end.
-
-Use as follows:
-
-call start()
-   do stuff
-call summary()
-
-Jim Tuccillo August 1999
-***************************************************************/
+/**
+ * @file
+ * @brief Make a system call to return various useful parameters.
+ *
+ * This code will make a system call to return various useful
+ * parameters.  When subroutine summary is called, a list of system
+ * resource statistics is printed to stdout.
+ *
+ * Users need to place a call to start() at the beginning of the
+ * section of code to be "measured" and a call to summary() at the
+ * end.
+ *
+ * Use as follows:
+ *
+ * @code
+ * call start()
+ *   do stuff
+ * call summary()
+ * @endcode
+ *
+ * @author Jim Tuccillo @date August 1999
+ */
 #include <stdio.h>
 #include <stdlib.h>
 #include <math.h>
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-//#include <sys/time.h>
+#include <sys/time.h>
 #include <sys/times.h>
 #include <sys/utsname.h>
 #ifdef _AIX
-#include <sys/proc.h>   
+#include <sys/proc.h>
 #endif
 #ifdef __linux__
 #include <errno.h>
 #include <sys/resource.h>
 #endif
-
-/* #include "trace_mpif.h" */
 
 static FILE *fp = NULL;
 int numtask, mypid;
@@ -46,18 +49,18 @@ double cpu_comm, wall_comm;
 extern double rtc ();
 #endif
 struct time_data {
-     double s_cpu;
-     double s_wall;
-     double f_cpu;
-     double f_wall;
-     double c_cpu;
-     double c_wall;
-     double c_bytes;
-     int    c_calls;
-     int    c_buckets[32];
-     float  c_sum[32];
-     double b_cpu[32];
-     double b_wall[32];
+    double s_cpu;
+    double s_wall;
+    double f_cpu;
+    double f_wall;
+    double c_cpu;
+    double c_wall;
+    double c_bytes;
+    int    c_calls;
+    int    c_buckets[32];
+    float  c_sum[32];
+    double b_cpu[32];
+    double b_wall[32];
 };
 
 struct time_data MPI_Allgather_data;
@@ -186,301 +189,283 @@ struct time_data MPI_Graph_neighbors_count_data;
 struct time_data MPI_Graphdims_get_data;
 struct time_data MPI_Topo_test_data;
 
-
+/**
+ * @param lng
+ * @author Jim Tuccillo @date August 1999
+ */
 int bucket (lng)
-   int    lng;
+int    lng;
 {
-   int    i, j;
-   if (lng <= 0) {return(0);}
-   for (i=1, j=--lng; j>0; ++i) {
-      j = j>>1;
-   }
-   return (i);
+    int    i, j;
+    if (lng <= 0) {return(0);}
+    for (i=1, j=--lng; j>0; ++i) {
+        j = j>>1;
+    }
+    return (i);
 }
 
-
-
+/**
+ * @param timer
+ * @author Jim Tuccillo @date August 1999
+ */
 void elapse (timer)
-     double *timer;
-
+double *timer;
 {
-
-/*
-
-      typedef struct { unsigned long tv_sec;
-               long tv_nsec; } timestruc;
-
-      timestruc TimePointer;
-      int ret;
-
-      ret = gettimer (TIMEOFDAY, &TimePointer);
-      if (ret != 0) {
-            printf ("getttimer FAILED!!!\n");
-            printf ("ret = %d\n", ret);
-            return;
-      }
-
-
-      *timer = ((double) TimePointer.tv_sec) + (((double) TimePointer.tv_nsec) * ((double) 0.000000001));
-      return;
-
-*/
 #ifdef _AIX
-      *timer = rtc();
+    *timer = rtc();
 #endif
 #ifdef __linux__
-  struct timeval st;
-  if (gettimeofday (&st, NULL) == -1) {
-    fprintf (stderr,
-             "elapse: gettimeofday: %s.\n",
-             strerror (errno));
-    *timer = 0.;
-  }
-  *timer = ((double) st.tv_sec) + 1.e-6 * ((double) st.tv_usec);
+    struct timeval st;
+    if (gettimeofday (&st, NULL) == -1) {
+        fprintf (stderr,
+                 "elapse: gettimeofday: %s.\n",
+                 strerror (errno));
+        *timer = 0.;
+    }
+    *timer = ((double) st.tv_sec) + 1.e-6 * ((double) st.tv_usec);
 #endif
-
 }
 
-
+/**
+ * @param usr
+ * @param sys
+ * @author Jim Tuccillo @date August 1999
+ */
 void cputim (usr, sys)
-     double *usr;
-     double *sys;
-
+double *usr;
+double *sys;
 {
+    double real;
+    typedef struct { int tms_utime;
+        int tms_stime;
+        int tms_cutime;
+        int tms_cstime; } tms;
 
-      double real;
-      typedef struct { int tms_utime;
-                       int tms_stime;
-                       int tms_cutime;
-                       int tms_cstime; } tms;
+    struct tms Time_buffer;
+    int ret;
 
-      struct tms Time_buffer;
-      int ret;
+    ret = times (&Time_buffer);
 
-      ret = times (&Time_buffer);
+    real = ((double) ret) * 0.01;
 
-      real = ((double) ret) * 0.01;
-
-      *usr = ((double) Time_buffer.tms_utime) * 0.01;
-      *sys = ((double) Time_buffer.tms_stime) * 0.01;
-      return;
-
+    *usr = ((double) Time_buffer.tms_utime) * 0.01;
+    *sys = ((double) Time_buffer.tms_stime) * 0.01;
+    return;
 }
 
-
+/**
+ * @param time
+ * @author Jim Tuccillo @date August 1999
+ */
 void start_timer (time)
-   struct time_data *time;
-
+struct time_data *time;
 {
-   double user, sys;
-   double wall;
+    double user, sys;
+    double wall;
 
-   cputim (&user, &sys);
-   elapse (&wall);
-   time->s_cpu = user + sys;
-   time->s_wall = wall;
+    cputim (&user, &sys);
+    elapse (&wall);
+    time->s_cpu = user + sys;
+    time->s_wall = wall;
 
-   return;
+    return;
 }
 
+/**
+ *
+ * @param time
+ * @author Jim Tuccillo @date August 1999
+ */
 void end_timer (time)
-   struct time_data *time;
-
+struct time_data *time;
 {
-   double user, sys;
-   double wall;
+    double user, sys;
+    double wall;
 
-   cputim (&user, &sys);
-   elapse (&wall);
-   time->f_cpu = user + sys;
-   time->f_wall = wall;
-   time->c_cpu += time->f_cpu - time->s_cpu;
-   time->c_wall += time->f_wall - time->s_wall;
+    cputim (&user, &sys);
+    elapse (&wall);
+    time->f_cpu = user + sys;
+    time->f_wall = wall;
+    time->c_cpu += time->f_cpu - time->s_cpu;
+    time->c_wall += time->f_wall - time->s_wall;
 
-   return;
+    return;
 }
 
-
-
-
+/**
+ *
+ * @author Jim Tuccillo @date August 1999
+ */
 void resource ()
 
 {
-
-      double usr, sys;
-      long data[14];
+    double usr, sys;
+    long data[14];
 #ifdef _AIX
-      typedef struct {
-          int             tv_sec;         /* seconds */
-          int             tv_usec;        /* microseconds */
-      } timeval;
+    typedef struct {
+        int             tv_sec;         /* seconds */
+        int             tv_usec;        /* microseconds */
+    } timeval;
 #endif
-      double user, system;
-      int ret;
+    double user, system;
+    int ret;
 
-      struct rusage RU;
-      ret = getrusage (0, &RU);
+    struct rusage RU;
+    ret = getrusage (0, &RU);
 
-      if (ret != 0) {
-            printf ("getrusage FAILED!!!\n");
-            printf ("ret = %d\n", ret);
-            return;
-      }
+    if (ret != 0) {
+        printf ("getrusage FAILED!!!\n");
+        printf ("ret = %d\n", ret);
+        return;
+    }
 
+    user = ((double) RU.ru_utime.tv_sec) + (((double) RU.ru_utime.tv_usec) * ((double) 0.000001));
+    system = ((double) RU.ru_stime.tv_sec) + (((double) RU.ru_stime.tv_usec) * ((double) 0.000001));
 
-      user = ((double) RU.ru_utime.tv_sec) + (((double) RU.ru_utime.tv_usec) * ((double) 0.000001));
-      system = ((double) RU.ru_stime.tv_sec) + (((double) RU.ru_stime.tv_usec) * ((double) 0.000001));
-
-
-     printf("*****************RESOURCE STATISTICS*******************************\n");
-     printf("The total amount of wall time                        = %f\n", tot_wall);
-     printf("The total amount of time in user mode                = %f\n", user);
-     printf("The total amount of time in sys mode                 = %f\n", system);
+    printf("*****************RESOURCE STATISTICS*******************************\n");
+    printf("The total amount of wall time                        = %f\n", tot_wall);
+    printf("The total amount of time in user mode                = %f\n", user);
+    printf("The total amount of time in sys mode                 = %f\n", system);
 #ifdef _AIX
-     printf("The maximum resident set size (KB)                   = %d\n", RU.ru_maxrss);
-     printf("Average shared memory use in text segment (KB*sec)   = %d\n", RU.ru_ixrss);
-     printf("Average unshared memory use in data segment (KB*sec) = %d\n", RU.ru_idrss);
-     printf("Average unshared memory use in stack segment(KB*sec) = %d\n", RU.ru_isrss);
-     printf("Number of page faults without I/O activity           = %d\n", RU.ru_minflt);
-     printf("Number of page faults with I/O activity              = %d\n", RU.ru_majflt);
-     printf("Number of times process was swapped out              = %d\n", RU.ru_nswap);
-     printf("Number of times filesystem performed INPUT           = %d\n", RU.ru_inblock);
-     printf("Number of times filesystem performed OUTPUT          = %d\n", RU.ru_oublock);
-     printf("Number of IPC messages sent                          = %d\n", RU.ru_msgsnd);
-     printf("Number of IPC messages received                      = %d\n", RU.ru_msgrcv);
-     printf("Number of Signals delivered                          = %d\n", RU.ru_nsignals);
-     printf("Number of Voluntary Context Switches                 = %d\n", RU.ru_nvcsw);
-     printf("Number of InVoluntary Context Switches               = %d\n", RU.ru_nivcsw);
+    printf("The maximum resident set size (KB)                   = %d\n", RU.ru_maxrss);
+    printf("Average shared memory use in text segment (KB*sec)   = %d\n", RU.ru_ixrss);
+    printf("Average unshared memory use in data segment (KB*sec) = %d\n", RU.ru_idrss);
+    printf("Average unshared memory use in stack segment(KB*sec) = %d\n", RU.ru_isrss);
+    printf("Number of page faults without I/O activity           = %d\n", RU.ru_minflt);
+    printf("Number of page faults with I/O activity              = %d\n", RU.ru_majflt);
+    printf("Number of times process was swapped out              = %d\n", RU.ru_nswap);
+    printf("Number of times filesystem performed INPUT           = %d\n", RU.ru_inblock);
+    printf("Number of times filesystem performed OUTPUT          = %d\n", RU.ru_oublock);
+    printf("Number of IPC messages sent                          = %d\n", RU.ru_msgsnd);
+    printf("Number of IPC messages received                      = %d\n", RU.ru_msgrcv);
+    printf("Number of Signals delivered                          = %d\n", RU.ru_nsignals);
+    printf("Number of Voluntary Context Switches                 = %d\n", RU.ru_nvcsw);
+    printf("Number of InVoluntary Context Switches               = %d\n", RU.ru_nivcsw);
 #endif
 #ifdef __linux__
-     printf ("The maximum resident set size (KB)                   = %ld\n", RU.ru_maxrss);
-     printf ("Number of page faults without I/O activity           = %ld\n", RU.ru_minflt);
-     printf ("Number of page faults with I/O activity              = %ld\n", RU.ru_majflt);
-     printf ("Number of times filesystem performed INPUT           = %ld\n", RU.ru_inblock);
-     printf ("Number of times filesystem performed OUTPUT          = %ld\n", RU.ru_oublock);
-     printf ("Number of Voluntary Context Switches                 = %ld\n", RU.ru_nvcsw);
-     printf ("Number of InVoluntary Context Switches               = %ld\n", RU.ru_nivcsw);
+    printf ("The maximum resident set size (KB)                   = %ld\n", RU.ru_maxrss);
+    printf ("Number of page faults without I/O activity           = %ld\n", RU.ru_minflt);
+    printf ("Number of page faults with I/O activity              = %ld\n", RU.ru_majflt);
+    printf ("Number of times filesystem performed INPUT           = %ld\n", RU.ru_inblock);
+    printf ("Number of times filesystem performed OUTPUT          = %ld\n", RU.ru_oublock);
+    printf ("Number of Voluntary Context Switches                 = %ld\n", RU.ru_nvcsw);
+    printf ("Number of InVoluntary Context Switches               = %ld\n", RU.ru_nivcsw);
 #endif
-     printf("*****************END OF RESOURCE STATISTICS*************************\n\n");
+    printf("*****************END OF RESOURCE STATISTICS*************************\n\n");
 
-
-      usr = user;
-      sys = system;
-      data[0] = RU.ru_maxrss;
-      data[1] = RU.ru_ixrss;
-      data[2] = RU.ru_idrss;
-      data[3] = RU.ru_isrss;
-      data[4] = RU.ru_minflt;
-      data[5] = RU.ru_majflt;
-      data[6] = RU.ru_nswap;
-      data[7] = RU.ru_inblock;
-      data[8] = RU.ru_oublock;
-      data[9] = RU.ru_msgsnd;
-      data[10] = RU.ru_msgrcv;
-      data[11] = RU.ru_nsignals;
-      data[12] = RU.ru_nvcsw;
-      data[13] = RU.ru_nivcsw;
-
-      return;
-
+    usr = user;
+    sys = system;
+    data[0] = RU.ru_maxrss;
+    data[1] = RU.ru_ixrss;
+    data[2] = RU.ru_idrss;
+    data[3] = RU.ru_isrss;
+    data[4] = RU.ru_minflt;
+    data[5] = RU.ru_majflt;
+    data[6] = RU.ru_nswap;
+    data[7] = RU.ru_inblock;
+    data[8] = RU.ru_oublock;
+    data[9] = RU.ru_msgsnd;
+    data[10] = RU.ru_msgrcv;
+    data[11] = RU.ru_nsignals;
+    data[12] = RU.ru_nvcsw;
+    data[13] = RU.ru_nivcsw;
+    return;
 }
 
-
-
-
-
+/**
+ *
+ * @param string
+ * @param time
+ * @author Jim Tuccillo @date August 1999
+ */
 void print_timing (string, time)
-   char             *string;
-   struct time_data *time;
-
+char             *string;
+struct time_data *time;
 {
+    if (time->c_calls > 0) {
+        fprintf (fp, "Information for %s: AVG. Length = %13.2f, CALLS = %d, WALL = %13.3f, CPU = %13.3f \n",
+                 string, (double) (time->c_bytes) / (double) time->c_calls, time->c_calls,
+                 time->c_wall, time->c_cpu);
+    }
 
+    if (time->c_wall > 0.001 ) {
+        fprintf (fp, "                %s: Total BYTES = %g, BW = %8.3f MBYTES/WALL SEC., BW = %8.3f MBYTES/CPU SEC.\n",
+                 string, time->c_bytes,
+                 ((double) time->c_bytes * 0.000001)/time->c_wall,
+                 ((double) time->c_bytes * 0.000001)/time->c_cpu);
+    }
 
-   if (time->c_calls > 0) {
-     fprintf (fp, "Information for %s: AVG. Length = %13.2f, CALLS = %d, WALL = %13.3f, CPU = %13.3f \n",
-     string, (double) (time->c_bytes) / (double) time->c_calls, time->c_calls,
-                       time->c_wall, time->c_cpu);
-   }
+    twall  += time->c_wall;
+    tcpu   += time->c_cpu;
+    tbytes += time->c_bytes * 0.000001;
 
-   if (time->c_wall > 0.001 ) {
-     fprintf (fp, "                %s: Total BYTES = %g, BW = %8.3f MBYTES/WALL SEC., BW = %8.3f MBYTES/CPU SEC.\n",
-     string, time->c_bytes,
-             ((double) time->c_bytes * 0.000001)/time->c_wall,
-             ((double) time->c_bytes * 0.000001)/time->c_cpu);
-   }
+    /* Print the distribution of the message lengths */
+    if (time->c_calls > 0) {
+        int i, j1, j2;
 
-   twall  += time->c_wall;
-   tcpu   += time->c_cpu;
-   tbytes += time->c_bytes * 0.000001;
+        j1 = 0; j2 = 0;
+        fprintf (fp, "       AVG. Length    # of Calls    MB/WALL Sec.  MB/CPU Sec.   WALL Secs.     CPU Secs.   \n");
+        if (time->c_buckets[0] >0) {
+            fprintf (fp, " %13.2f %13d    %13.3f %13.3f %13.4f %13.4f \n",
+                     time->c_sum[0]/(float)time->c_buckets[0], time->c_buckets[0],
+                     ((double) time->c_sum[0] * 0.000001)/time->b_wall[0],
+                     ((double) time->c_sum[0] * 0.000001)/time->b_cpu[0],
+                     time->b_wall[0], time->b_cpu[0]);
+        }
+        time->c_buckets[3] = time->c_buckets[1] + time->c_buckets[2] + time->c_buckets[3];
+        j1 = 1; j2 = 4;
+        for (i =3; i < 31; ++i) {
+            if (time->c_buckets[i] > 0) {
+                fprintf (fp, " %13.2f %13d    %13.3f %13.3f %13.4f %13.4f \n",
+                         time->c_sum[i]/(float)time->c_buckets[i], time->c_buckets[i],
+                         ((double) time->c_sum[i] * 0.000001)/time->b_wall[i],
+                         ((double) time->c_sum[i] * 0.000001)/time->b_cpu[i],
+                         time->b_wall[i], time->b_cpu[i]);
+            }
+            j1 = j2 +1;
+            j2 = j2 + j2;
+        }
 
-   /* Print the distribution of the message lengths */
-   if (time->c_calls > 0) {
-     int i, j1, j2;
-
-     j1 = 0; j2 = 0;
-     fprintf (fp, "       AVG. Length    # of Calls    MB/WALL Sec.  MB/CPU Sec.   WALL Secs.     CPU Secs.   \n");
-     if (time->c_buckets[0] >0) {
-      fprintf (fp, " %13.2f %13d    %13.3f %13.3f %13.4f %13.4f \n",
-                   time->c_sum[0]/(float)time->c_buckets[0], time->c_buckets[0],
-                   ((double) time->c_sum[0] * 0.000001)/time->b_wall[0],
-                   ((double) time->c_sum[0] * 0.000001)/time->b_cpu[0],
-                   time->b_wall[0], time->b_cpu[0]);
-       }
-     time->c_buckets[3] = time->c_buckets[1] + time->c_buckets[2] + time->c_buckets[3];
-     j1 = 1; j2 = 4;
-     for (i =3; i < 31; ++i) {
-      if (time->c_buckets[i] > 0) {
-      fprintf (fp, " %13.2f %13d    %13.3f %13.3f %13.4f %13.4f \n",
-                   time->c_sum[i]/(float)time->c_buckets[i], time->c_buckets[i],
-                   ((double) time->c_sum[i] * 0.000001)/time->b_wall[i],
-                   ((double) time->c_sum[i] * 0.000001)/time->b_cpu[i],
-                   time->b_wall[i], time->b_cpu[i]);
-      }
-       j1 = j2 +1;
-       j2 = j2 + j2;
-     }
-
-   fprintf (fp, "\n");
-
-   }
-
+        fprintf (fp, "\n");
+    }
 }
 
+/**
+ * @param returnVal
+ * @author Jim Tuccillo @date August 1999
+ */
 void summary_ (int *returnVal)
 {
-
-  double temp, temp1;
-  char trace_file[255], processor[8];
+    double temp, temp1;
+    char trace_file[255], processor[8];
 
 /*
-    MPI_Finalize - prototyping replacement for MPI_Finalize
+  MPI_Finalize - prototyping replacement for MPI_Finalize
 */
-
-   elapse(&final_wall);
-   tot_wall = final_wall - start_wall;
-
+    elapse(&final_wall);
+    tot_wall = final_wall - start_wall;
 
     resource();
 
-   if (fp) fclose (fp);
-
-  return;
+    if (fp) fclose (fp);
+    return;
 }
 
+/**
+ *
+ * @author Jim Tuccillo @date August 1999
+ */
 void start_ ()
 {
-  int stateid;
-  int  Argc;
-  char **Argv;
+    int stateid;
+    int  Argc;
+    char **Argv;
 
-  char *answer;
- 
+    char *answer;
 
-  trace_flag=1;
+    trace_flag=1;
 
-  profile = 0;
-  elapse (&start_wall);
-
-  return;
+    profile = 0;
+    elapse (&start_wall);
+    return;
 }
-

--- a/src/summary.c
+++ b/src/summary.c
@@ -3,7 +3,7 @@
  * @brief Make a system call to return various useful parameters.
  *
  * This code will make a system call to return various useful
- * parameters.  When subroutine summary is called, a list of system
+ * parameters. When subroutine summary() is called, a list of system
  * resource statistics is printed to stdout.
  *
  * Users need to place a call to start() at the beginning of the

--- a/tests/test_summary.f90
+++ b/tests/test_summary.f90
@@ -1,3 +1,8 @@
+! This is a test in the NCEPLIBS-w3emc project.
+!
+! Test the summary subroutine.
+!
+! Kyle Gerheiser
 program test_summary
   implicit none
   print *, "Testing summary()..."

--- a/tests/test_w3fi71.f90
+++ b/tests/test_w3fi71.f90
@@ -1,3 +1,8 @@
+! This is a test in the NCEPLIBS-w3emc project.
+!
+! Test the w3fi71() subroutine.
+!
+! George Gayno
  program test_w3fi71
 
 ! Call routine w3fi71 for NCEP grids 172 and 220.

--- a/tests/test_w3tagb.f90
+++ b/tests/test_w3tagb.f90
@@ -1,3 +1,8 @@
+! This is a test in the NCEPLIBS-w3emc project.
+!
+! Test the w3tagb function.
+!
+! Kyle Gerheiser
 program test_w3tagb
   implicit none
   integer :: year, julian_day, hundreths_of_julian_day

